File: ios/SpartanKit/Sources/SpartanKit/ProjectionEngine.swift
import Foundation

/// One projected range at a week offset from today. `low`/`high` are *numeric* bounds
/// (`low <= high` always); whether the top of the range is the desirable end comes from
/// `MetricProjection.higherIsBetter`, not from this struct.
public struct ProjectionPoint: Codable, Equatable {
    public let week: Int
    public let low: Double
    public let high: Double

    public init(week: Int, low: Double, high: Double) {
        self.week = week
        self.low = low
        self.high = high
    }
}

/// An illustrative 8-week range for one WHOOP metric, produced by `ProjectionEngine`.
/// Every user-facing string on this type has passed the `SafetyEngine` blocked-phrase check.
public struct MetricProjection: Codable, Equatable {
    public let metric: MetricType
    public let label: String
    public let unit: String
    public let currentValue: Double
    public let points: [ProjectionPoint]
    public let assumption: String
    public let higherIsBetter: Bool

    public init(
        metric: MetricType,
        label: String,
        unit: String,
        currentValue: Double,
        points: [ProjectionPoint],
        assumption: String,
        higherIsBetter: Bool
    ) {
        self.metric = metric
        self.label = label
        self.unit = unit
        self.currentValue = currentValue
        self.points = points
        self.assumption = assumption
        self.higherIsBetter = higherIsBetter
    }
}

/// Deterministic, rules-based "where could this metric plausibly go" projector.
/// iOS port of the Android `ProjectionEngine`: same tiers, same deltas, same clamps.
///
/// Modeling rules (all illustrative wellness estimates — never medical predictions):
/// - Consistency tier scales the full 8-week effect: >=5 plan days/week -> full effect,
///   3-4 -> 0.65x, 1-2 -> 0.35x, 0 -> flat (every week holds at today's value).
/// - Resting heart rate: -3..-6 bpm at week 8 (scaled by tier), floored at
///   `max(45, current * 0.85)` and never pushed above today's value by the floor.
/// - HRV: +5%..+15% at week 8 (scaled by tier), capped at `current * 1.25`.
/// - Recovery: +5..+12 points at week 8 (scaled by tier), capped at 90.
/// - Points at weeks [0, 2, 4, 6, 8], linearly ramped; week 0 equals the current value.
/// - Values rounded to 1 decimal; numeric `low <= high` on every point.
public final class ProjectionEngine {

    /// Shown alongside every projection surface. Keep in sync with Android's
    /// `ProjectionEngine.DISCLAIMER`.
    // Kept byte-identical to the Android ProjectionEngine.DISCLAIMER for cross-platform parity.
    public static let disclaimer =
        "Typical ranges seen with consistent aerobic and strength training in general research" +
        " — not a prediction or a guarantee for you. Individual results vary." +
        " This is wellness guidance, not medical advice."

    private static let weeks: [Int] = [0, 2, 4, 6, 8]
    private static let horizonWeeks = 8.0

    private let safety = SafetyEngine()

    public init() {
        // The disclaimer is user-facing copy: hold it to the same guardrail as everything else.
        safety.sanitizeOrFatal(Self.disclaimer)
    }

    /// Projects each metric whose current value is provided. Output order:
    /// resting heart rate, HRV, recovery. Same inputs always produce the same output.
    public func project(
        restingHeartRate: Double?,
        hrvMs: Double?,
        recoveryScore: Int?,
        consistencyDays7: Int
    ) -> [MetricProjection] {
        let factor = Self.consistencyFactor(consistencyDays7)
        let tierPhrase = Self.tierPhrase(consistencyDays7)
        var projections: [MetricProjection] = []

        if let rhr = restingHeartRate {
            // Floor never exceeds today's value, so an already-low RHR just holds flat.
            let floorValue = min(max(45.0, rhr * 0.85), rhr)
            let points = Self.weeks.map { week -> ProjectionPoint in
                let ramp = factor * Double(week) / Self.horizonWeeks
                let bestCase = max(rhr - 6.0 * ramp, floorValue)
                let modestCase = max(rhr - 3.0 * ramp, floorValue)
                return Self.point(week: week, bestCase, modestCase)
            }
            projections.append(makeProjection(
                metric: .restingHeartRate,
                label: "Resting heart rate",
                unit: "bpm",
                currentValue: rhr,
                points: points,
                tierPhrase: tierPhrase,
                mechanism: "Steady aerobic activity and good sleep tend to lower resting heart rate a little at a time.",
                higherIsBetter: false
            ))
        }

        if let hrv = hrvMs {
            let cap = hrv * 1.25
            let points = Self.weeks.map { week -> ProjectionPoint in
                let ramp = factor * Double(week) / Self.horizonWeeks
                let modestCase = min(hrv * (1.0 + 0.05 * ramp), cap)
                let bestCase = min(hrv * (1.0 + 0.15 * ramp), cap)
                return Self.point(week: week, modestCase, bestCase)
            }
            projections.append(makeProjection(
                metric: .hrvRmssd,
                label: "HRV",
                unit: "ms",
                currentValue: hrv,
                points: points,
                tierPhrase: tierPhrase,
                mechanism: "Regular movement, steady sleep, and recovery habits tend to lift HRV toward the upper end of your own range.",
                higherIsBetter: true
            ))
        }

        if let recovery = recoveryScore {
            let current = Double(recovery)
            // Cap never drops below today's value, so a score already above 90 holds flat.
            let ceiling = max(90.0, current)
            let points = Self.weeks.map { week -> ProjectionPoint in
                let ramp = factor * Double(week) / Self.horizonWeeks
                let modestCase = min(current + 5.0 * ramp, ceiling)
                let bestCase = min(current + 12.0 * ramp, ceiling)
                return Self.point(week: week, modestCase, bestCase)
            }
            projections.append(makeProjection(
                metric: .recoveryScore,
                label: "Recovery",
                unit: "%",
                currentValue: current,
                points: points,
                tierPhrase: tierPhrase,
                mechanism: "Balanced load and better sleep usually show up as a higher recovery score over weeks.",
                higherIsBetter: true
            ))
        }

        return projections
    }

    // MARK: - Internals

    /// >=5 days -> full effect; 3-4 -> 0.65; 1-2 -> 0.35; 0 (or less) -> flat.
    private static func consistencyFactor(_ days: Int) -> Double {
        switch days {
        case 5...: return 1.0
        case 3...4: return 0.65
        case 1...2: return 0.35
        default: return 0.0
        }
    }

    private static func tierPhrase(_ days: Int) -> String {
        switch days {
        case 5...: return "you keep up 5 or more plan days a week"
        case 3...4: return "you keep up 3 to 4 plan days a week"
        case 1...2: return "you keep up 1 to 2 plan days a week"
        default: return "consistency stays at zero plan days, so the range holds flat at today's value"
        }
    }

    /// Orders the two candidate values numerically and rounds both to 1 decimal.
    private static func point(week: Int, _ a: Double, _ b: Double) -> ProjectionPoint {
        ProjectionPoint(week: week, low: round1(min(a, b)), high: round1(max(a, b)))
    }

    private static func round1(_ value: Double) -> Double {
        (value * 10.0).rounded() / 10.0
    }

    private func makeProjection(
        metric: MetricType,
        label: String,
        unit: String,
        currentValue: Double,
        points: [ProjectionPoint],
        tierPhrase: String,
        mechanism: String,
        higherIsBetter: Bool
    ) -> MetricProjection {
        let assumption =
            "Illustrative estimate, not a prediction or a guarantee. Assumes \(tierPhrase)" +
            " for the next 8 weeks. \(mechanism)"
        return MetricProjection(
            metric: metric,
            label: safety.sanitizeOrFatal(label),
            unit: unit,
            currentValue: currentValue,
            points: points,
            assumption: safety.sanitizeOrFatal(assumption),
            higherIsBetter: higherIsBetter
        )
    }
}

File: app/src/main/java/com/spartan/domain/engine/ProjectionEngine.kt
package com.spartan.domain.engine

import com.spartan.domain.model.MetricType
import kotlin.math.ceil
import kotlin.math.floor
import kotlin.math.round

/**
 * One projected point on the 8-week improvement chart.
 *
 * [low] and [high] are the numeric bounds of the projected VALUE, always ordered
 * `low <= high` — they are NOT "pessimistic/optimistic" in improvement terms. For a
 * lower-is-better metric like resting heart rate the numerically low bound is the
 * more-improved end of the range; for higher-is-better metrics it is the less-improved end.
 */
data class ProjectionPoint(val week: Int, val low: Double, val high: Double)

data class MetricProjection(
    val metric: MetricType,
    val label: String,            // e.g. "Resting heart rate"
    val unit: String,             // e.g. "bpm"
    val currentValue: Double,
    val points: List<ProjectionPoint>,   // weeks 0,2,4,6,8 — week 0 == current (low==high==current)
    val assumption: String,       // e.g. "If you complete your plan on 5+ days a week"
    val higherIsBetter: Boolean,
)

/**
 * Projects an honest EXPECTED IMPROVEMENT RANGE for a user's wearable metrics over the
 * next 8 weeks if they keep completing their Spartan plan.
 *
 * The ranges are typical effects of consistent aerobic-base plus strength training reported
 * in general exercise-science findings, scaled down conservatively when the user's trailing
 * 7-day consistency is low. They are population-typical ranges — never a personal prediction,
 * diagnosis, or guarantee. Every surface that renders these projections must also render
 * [DISCLAIMER]. The model is fully deterministic: same inputs, same output.
 */
class ProjectionEngine(private val safetyEngine: SafetyEngine = SafetyEngine()) {

    init {
        // Fail fast at construction if the always-rendered copy ever drifts into blocked phrasing.
        safetyEngine.sanitize(DISCLAIMER)
        safetyEngine.sanitize(ASSUMPTION_FULL)
        safetyEngine.sanitize(ASSUMPTION_INACTIVE)
    }

    /**
     * Consistency = days with a completed activity in the trailing 7.
     * Returns projections only for metrics with a current value.
     */
    fun project(
        restingHeartRate: Double?,
        hrvMs: Double?,
        recoveryScore: Int?,
        consistencyDays7: Int,
    ): List<MetricProjection> {
        val days = consistencyDays7.coerceIn(0, 7)
        val effect = when {
            days >= FULL_EFFECT_MIN_DAYS -> EFFECT_FULL
            days >= 3 -> EFFECT_MODERATE
            days >= 1 -> EFFECT_LIGHT
            else -> 0.0
        }
        val assumption = safetyEngine.sanitize(
            when {
                days >= FULL_EFFECT_MIN_DAYS -> ASSUMPTION_FULL
                days >= 1 -> "At your current consistency of $days days a week"
                else -> ASSUMPTION_INACTIVE
            },
        )
        val projections = mutableListOf<MetricProjection>()
        restingHeartRate?.let { projections += projectRestingHeartRate(it, effect, assumption) }
        hrvMs?.let { projections += projectHrv(it, effect, assumption) }
        recoveryScore?.let { projections += projectRecovery(it.toDouble(), effect, assumption) }
        return projections
    }

    /**
     * Resting heart rate (lower is better): full-effect improvement of 3.0–6.0 bpm at week 8,
     * ramping linearly, never projected below max(45 bpm, 85% of current) and never above current.
     */
    private fun projectRestingHeartRate(
        current: Double,
        effect: Double,
        assumption: String,
    ): MetricProjection {
        // Ceil the floor onto the 1-decimal grid so rounding can never land below it.
        val floorValue = ceilTo1(maxOf(RHR_FLOOR_BPM, current * RHR_FLOOR_FRACTION))
        return buildProjection(
            metric = MetricType.RESTING_HEART_RATE,
            current = current,
            effect = effect,
            assumption = assumption,
            // The biggest bpm drop is the most-improved bound, i.e. the numerically LOW value.
            rawLow = { f -> current - RHR_DROP_MAX_BPM * effect * f },
            rawHigh = { f -> current - RHR_DROP_MIN_BPM * effect * f },
            clamp = { v -> v.coerceAtLeast(floorValue).coerceAtMost(current) },
        )
    }

    /**
     * HRV (higher is better): full-effect improvement of +5% to +15% at week 8,
     * ramping linearly, capped at 125% of current.
     */
    private fun projectHrv(current: Double, effect: Double, assumption: String): MetricProjection {
        // Floor the cap onto the 1-decimal grid so rounding can never land above it.
        val cap = floorTo1(current * HRV_CAP_FRACTION)
        return buildProjection(
            metric = MetricType.HRV_RMSSD,
            current = current,
            effect = effect,
            assumption = assumption,
            rawLow = { f -> current * (1.0 + HRV_GAIN_MIN_FRACTION * effect * f) },
            rawHigh = { f -> current * (1.0 + HRV_GAIN_MAX_FRACTION * effect * f) },
            clamp = { v -> v.coerceAtMost(cap) },
        )
    }

    /**
     * Recovery score (higher is better): full-effect improvement of +5 to +12 points at week 8,
     * ramping linearly. Projected gains are hard-capped at 90 and never exceed 99; a user already
     * at or above the cap is shown a flat line at their current value (we never project a decline).
     */
    private fun projectRecovery(current: Double, effect: Double, assumption: String): MetricProjection {
        val cap = minOf(RECOVERY_HARD_LIMIT, maxOf(RECOVERY_SOFT_CAP, current))
        return buildProjection(
            metric = MetricType.RECOVERY_SCORE,
            current = current,
            effect = effect,
            assumption = assumption,
            rawLow = { f -> current + RECOVERY_GAIN_MIN_POINTS * effect * f },
            rawHigh = { f -> current + RECOVERY_GAIN_MAX_POINTS * effect * f },
            clamp = { v -> v.coerceAtMost(cap).coerceAtLeast(current) },
        )
    }

    /**
     * Shared point builder. [rawLow]/[rawHigh] take the linear ramp fraction (week / 8) and
     * must satisfy rawLow(f) <= rawHigh(f); [clamp] must be monotone non-decreasing, so the
     * numeric ordering low <= high survives clamping and rounding.
     */
    private fun buildProjection(
        metric: MetricType,
        current: Double,
        effect: Double,
        assumption: String,
        rawLow: (Double) -> Double,
        rawHigh: (Double) -> Double,
        clamp: (Double) -> Double,
    ): MetricProjection {
        val points = PROJECTED_WEEKS.map { week ->
            if (week == 0 || effect == 0.0) {
                // Week 0 is always exactly the current value; zero consistency projects flat.
                ProjectionPoint(week, current, current)
            } else {
                val fraction = week / HORIZON_WEEKS
                ProjectionPoint(
                    week = week,
                    low = finish(rawLow(fraction), clamp, current),
                    high = finish(rawHigh(fraction), clamp, current),
                )
            }
        }
        return MetricProjection(
            metric = metric,
            label = metric.label,
            unit = metric.unit,
            currentValue = current,
            points = points,
            assumption = assumption,
            higherIsBetter = !metric.lowerIsBetter,
        )
    }

    /** Clamp then round to 1 decimal; a bound clamped back to the current value stays exact. */
    private fun finish(raw: Double, clamp: (Double) -> Double, current: Double): Double {
        val clamped = clamp(raw)
        return if (clamped == current) current else roundTo1(clamped)
    }

    private fun roundTo1(value: Double): Double = round(value * 10.0) / 10.0

    private fun ceilTo1(value: Double): Double = ceil(value * 10.0) / 10.0

    private fun floorTo1(value: Double): Double = floor(value * 10.0) / 10.0

    companion object {
        /** Shown wherever projections render. */
        const val DISCLAIMER: String = "Typical ranges seen with consistent aerobic and strength training in general research — not a prediction or a guarantee for you. Individual results vary. This is wellness guidance, not medical advice."

        private const val ASSUMPTION_FULL = "If you complete your plan on 5+ days a week"
        private const val ASSUMPTION_INACTIVE =
            "Complete a few activities each week to see a projected range"

        private val PROJECTED_WEEKS = listOf(0, 2, 4, 6, 8)
        private const val HORIZON_WEEKS = 8.0

        private const val FULL_EFFECT_MIN_DAYS = 5
        private const val EFFECT_FULL = 1.0
        private const val EFFECT_MODERATE = 0.65
        private const val EFFECT_LIGHT = 0.35

        private const val RHR_DROP_MIN_BPM = 3.0
        private const val RHR_DROP_MAX_BPM = 6.0
        private const val RHR_FLOOR_BPM = 45.0
        private const val RHR_FLOOR_FRACTION = 0.85

        private const val HRV_GAIN_MIN_FRACTION = 0.05
        private const val HRV_GAIN_MAX_FRACTION = 0.15
        private const val HRV_CAP_FRACTION = 1.25

        private const val RECOVERY_GAIN_MIN_POINTS = 5.0
        private const val RECOVERY_GAIN_MAX_POINTS = 12.0
        private const val RECOVERY_SOFT_CAP = 90.0
        private const val RECOVERY_HARD_LIMIT = 99.0
    }
}
